sql/src/main/java/io/druid/sql/calcite/schema/SystemSchema.java

```diff
@@ -0,0 +1,435 @@
+        DruidLeaderClient druidLeaderClient,
+        ObjectMapper jsonMapper
+    )
+    {
+      try {
+        FullResponseHolder response = druidLeaderClient.go(
+            druidLeaderClient.makeRequest(
+                HttpMethod.GET,
+                StringUtils.format(
+                    "/druid/coordinator/v1/datasources?full"
+                )
+            )
+        );
+
+        if (!response.getStatus().equals(HttpResponseStatus.OK)) {
+          throw new ISE(
+              "Error while fetching metadata segments status[%s] content[%s]",
+              response.getStatus(),
+              response.getContent()
+          );
+        }
+        return jsonMapper.readValue(
+            response.getContent(), new TypeReference<List<ImmutableDruidDataSource>>()
+            {
+            }
+        );
+      }
+      catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+    }
+  }
+
+  private class ServersTable extends AbstractTable implements ScannableTable
+  {
+
+    @Override
+    public RelDataType getRowType(RelDataTypeFactory typeFactory)
+    {
+      return SERVERS_SIGNATURE.getRelDataType(typeFactory);
+    }
+
+    @Override
+    public TableType getJdbcTableType()
+    {
+      return TableType.SYSTEM_TABLE;
+    }
+
+    @Override
+    public Enumerable<Object[]> scan(DataContext root)
+    {
+      Map<String, QueryableDruidServer> serverViewClients = serverView.getClients();
+      final FluentIterable<Object[]> results = FluentIterable
+          .from(serverViewClients.values())
+          .transform(val -> new Object[]{
+              val.getServer().getHost(),
+              val.getServer().getType(),
+              val.getServer().getTier(),
+              val.getServer().getCurrSize(),
+              val.getServer().getMaxSize()
+          });
+      return Linq4j.asEnumerable(results);
+    }
+  }
+
+  private class ServerSegmentsTable extends AbstractTable implements ScannableTable
+  {
+
+    @Override
+    public RelDataType getRowType(RelDataTypeFactory typeFactory)
+    {
+      return SERVERSEGMENTS_SIGNATURE.getRelDataType(typeFactory);
+    }
+
+    @Override
+    public TableType getJdbcTableType()
+    {
+      return TableType.SYSTEM_TABLE;
+    }
+
+    @Override
+    public Enumerable<Object[]> scan(DataContext root)
+    {
+      final List<Object[]> rows = new ArrayList<>();
+      final Map<String, QueryableDruidServer> serverViewClients = serverView.getClients();
+      for (QueryableDruidServer queryableDruidServer : serverViewClients.values()) {
+        Object[] row = new Object[SERVERSEGMENTS_TABLE_SIZE];
+        final DruidServer druidServer = queryableDruidServer.getServer();
+        final Map<String, DataSegment> segmentMap = druidServer.getSegments();
+        for (DataSegment segment : segmentMap.values()) {
+          row[0] = druidServer.getHost();
+          row[1] = segment.getIdentifier();
+          rows.add(row);
```

Comment:
This has the same issue as the segments table above. It doesn't work to reuse arrays like this.

Reply:
moved `Object[] row = new Object[SERVERSEGMENTS_TABLE_SIZE];` inside the inner loop